torch/nn/modules/lazy.py

```diff
@@ -0,0 +1,246 @@
+    ...
+    ...    def forward(self, input):
+    ...        x = self.relu1(self.fc1(input))
+    ...        y = self.relu2(self.fc2(x))
+    ...        return y
+    >>> # constructs a network with lazy modules
+    >>> lazy_mlp = LazyMLP()
+    >>> # transforms the network's device and dtype
+    >>> # NOTE: these transforms can and should be applied after construction and before any 'dry runs'
+    >>> lazy_mlp = mlp.cuda().double()
+    >>> lazy_mlp
+    LazyMLP(
+      (fc1): LazyLinear(in_features=0, out_features=10, bias=True)
+      (relu1): ReLU()
+      (fc2): LazyLinear(in_features=0, out_features=1, bias=True)
+      (relu2): ReLU()
+    )
+    >>> # performs a dry run to initialize the network's lazy modules
+    >>> lazy_mlp(torch.ones(10,10).cuda())
+    >>> # after initialization, LazyLinear modules become regular Linear modules
+    >>> lazy_mlp
+    LazyMLP(
+      (fc1): Linear(in_features=10, out_features=10, bias=True)
+      (relu1): ReLU()
+      (fc2): Linear(in_features=10, out_features=1, bias=True)
+      (relu2): ReLU()
+    )
+    >>> # attaches an optimizer, since parameters can now be used as usual
+    >>> optim = torch.optim.SGD(mlp.parameters(), lr=0.01)
+
+    Note that lazy modules cannot validate if the shape is correct during
+    deserialization when loading an initialized parameter into an uninitialized one.
+    Also note that the initial weights values may become different
+    from the original (non-lazy) module even if the random seed is manually
+    configured, as the order of initialization is different from the original
+    one.
+
+    In the example above, if both layers are `LazyLinear` the initialization
+    order is equivalent to the non-lazy one, so weights would be the same if the same
+    random seed is used. However, if only the first layer is lazy, then the initialization
+    of the weights happens in different order. The second layer at creation time, and the
+    first layer is initialized in the dry run, resulting in different weights than the non
+    lazy version of the network.
+
+    The default value of lazy parameters are ``UninitializedParameter()``.
+
+    .. note:: A `LazyModule`, or a module containing a `LazyModule` requires an explicit
+    dummy forward call in the following scenarios.
+
+    * Access an :class:`UnitializedParameter` data or properties
+    >>> lazy_module.weight.shape
+
+    * Use an Optimizer that accesses Parameters data in their constructor 
+    >>> optim = torch.optim.AdaGrad(lazy_module.parameters())
+
+    * Use :class:`torch.nn.DataParallel` or :class:`torch.nn.DistributedDataParallel`
+    >>> module_dp = torch.nn.DataParallel(lazy_module)
+
+    * Module reparametrization using :func:`torch.nn.utils.weight_norm` or :func:`torch.nn.utils.spectral_norm`
+    >>> torch.nn.utils.weight_norm(lazy_module)
+
+    * When dealing with non-default module initialization
+    >>> torch.nn.init.uniform_(lazy_module.weight)
+
+    .. warning:
```

Comment:
I suggest moving this warning up in the documentation (see my updated comment about the start of these docs).